internal/pkg/gc/actions.go

```diff
@@ -0,0 +1,134 @@
+	"math/rand"
+	"net/http"
+	"time"
+
+	"github.com/rs/zerolog/log"
+
+	"github.com/elastic/fleet-server/v7/internal/pkg/bulk"
+	"github.com/elastic/fleet-server/v7/internal/pkg/dl"
+	"github.com/elastic/fleet-server/v7/internal/pkg/es"
+	"github.com/elastic/fleet-server/v7/internal/pkg/scheduler"
+	"github.com/elastic/fleet-server/v7/internal/pkg/wait"
+)
+
+const (
+	defaultMaxWaitInCleanupLoop         = 10 * time.Second // the wait in the cleanup loop between iteration is random
+	defaultActionsSelectSize            = 1000
+	defaultActionsCleanupBeforeInterval = 24 * 30 * time.Hour
+)
+
+type ActionsCleanupConfig struct {
+	maxWaitInCleanupLoop  time.Duration
+	actionsSelectSize     int
+	cleanupBeforeInterval time.Duration
+}
+
+type ActionsCleanupOpt func(c *ActionsCleanupConfig)
+
+func WithMaxWaitInCleanupLoop(maxWaitInCleanupLoop time.Duration) ActionsCleanupOpt {
+	return func(c *ActionsCleanupConfig) {
+		c.maxWaitInCleanupLoop = maxWaitInCleanupLoop
+	}
+}
+
+func WithActionSelectSize(actionsSelectSize int) ActionsCleanupOpt {
+	return func(c *ActionsCleanupConfig) {
+		c.actionsSelectSize = actionsSelectSize
+	}
+}
+
+func WithActionCleanupBeforeInterval(cleanupBeforeInterval time.Duration) ActionsCleanupOpt {
+	return func(c *ActionsCleanupConfig) {
+		c.cleanupBeforeInterval = cleanupBeforeInterval
+	}
+}
+
+func getActionsGCFunc(bulker bulk.Bulk, cleanupBeforeInterval time.Duration) scheduler.WorkFunc {
+	return func(ctx context.Context) error {
+		return cleanupActions(ctx, dl.FleetActions, bulker,
+			WithActionCleanupBeforeInterval(cleanupBeforeInterval))
+	}
+}
+
+func cleanupActions(ctx context.Context, index string, bulker bulk.Bulk, opts ...ActionsCleanupOpt) error {
+	log := log.With().Str("ctx", "fleet actions cleanup").Logger()
+
+	c := ActionsCleanupConfig{
+		cleanupBeforeInterval: defaultActionsCleanupBeforeInterval,
+		actionsSelectSize:     defaultActionsSelectSize,
+		maxWaitInCleanupLoop:  defaultMaxWaitInCleanupLoop,
+	}
+
+	for _, opt := range opts {
+		opt(&c)
+	}
+
+	// Cleanup expired actions where expired timetamp is older than current time minus cleanupBeforeInterval
+	// Example: cleanup up actions that expired more than two weeks ago
+	expiredBefore := time.Now().Add(-c.cleanupBeforeInterval)
+
+	// Random generator for calculating random pause duration in the cleanup loop
+	r := rand.New(rand.NewSource(time.Now().Unix()))
+
+	var (
+		hits []es.HitT
+		err  error
+	)
+
+	for {
+		log.Debug().Str("expired_before", expiredBefore.UTC().Format(time.RFC3339)).Msgf("find actions that expired before given date/time")
+		hits, err = dl.FindExpiredActionsHitsForIndex(ctx, index, bulker, expiredBefore, c.actionsSelectSize)
+		if err != nil {
+			return err
+		}
+
+		if len(hits) == 0 {
+			log.Debug().Msg("no more expired actions found, done cleaning")
+			return nil
+		}
+
+		log.Debug().Int("count", len(hits)).Msg("delete expired actions")
+		for _, hit := range hits {
+			err = bulker.Delete(ctx, index, hit.Id)
```

Comment:
multi delete? https://github.com/elastic/fleet-server/blob/master/internal/pkg/bulk/opMulti.go#L25

Reply:
Updated to use MDelete